detect user libraries and add to shared.R path if needed

gordonwoodhull opened the pull request:
@s-u, please review. I tested the code in a notebook but I need a multiuser setup in order to test it in reality.

My summary of our earlier conversation is here:
https://github.com/att/rcloud.htmlwidgets/pull/17#issuecomment-308551464

Unfortunately we can't use `getNamespaceInfo` here because most of the time we'll be given the complete path and need to determine the package from it. That's why this code compares path prefixes across `.libPaths()` instead.

gordonwoodhull committed:
detect user libraries and add to shared.R path if needed

s-u reviewed (commented):
Um, I still don't understand why it's not using `system.file(file, package=pkg)` .. that would be so much easier. Also the current code won't work is the notebook used `library(foo, lib.loc=...)` since it will pick the wrong library. I'm nto 100% sure about the detection of the user library  - perhaps use `([^/]+)` since I'm not sure I can specify what a valid username is. Finally, we may want to warn if the  file is inaccessible, i.e. if is was loaded from anything else than the supported two libraries and/or if the permissions fail (if we have a way to convey errors...).

gordonwoodhull commented:
Unfortunately we only get the (full) path, not the package name, most of the time. It's the dependency format we get from htmltools. package+relative path seems to be a recent innovation there.

<strike>Seems to me this code *will* work for `lib.loc` libraries in the case where we get the full path.</strike> In the case where we get package + relative path, maybe that is the right time to use `getNamespaceInfo`, essentially to find out how the package was loaded.

Yeah, sure.... error handling. Error handling is for wimps.

gordonwoodhull committed:
address @s-u feedback

* if package & path, use getNamespaceInfo to seek the package path,
in case it was loaded with lib.loc
* if full path, check it against user lib first before .libPaths()
* print a warning if the subdirectory is not supported
* simpler username regex

Currently does no check if the file exists or permissions are ok - it's
just string manipulation against various library paths. We could add that
in a later commit.

gordonwoodhull commented:
I've rewritten it but not tested yet - will test tonight.

gordonwoodhull committed:
version and changelog

gordonwoodhull committed:
don't grab rest of path as user using gsub

s-u commented:
I think my main comment still stands - the original code was written by someone who was replicating `system.file()` in very complicated ways. `system.file()` does all of the above - it uses the namespaces first then looks on the lib path and then also check for existence. It should be possible to simply replace all this with the code in `shared.R` plus the user lib detection.

gordonwoodhull commented:
This code looks at absolute file system paths and produces relative paths suitable for `shared.R`

It is the inverse operation from `system.file`.

Yes there is one case where we know the package and relative path already, and that case is very simple. I don't see how `system.file` helps in the other case.

gordonwoodhull commented:
I've already tested and merged this for 1.0.8 and 1.1.1.

If you want to rewrite for 1.1.2 be my guest.
